File: sdk-go/images_test.go
package boxkite

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"testing"
)

func TestCreateImage_SendsExactRequestBody(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/images" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["base"] != "boxkite-minimal" {
			t.Errorf("unexpected base: %v", body["base"])
		}
		pythonPackages, _ := body["python_packages"].([]any)
		if len(pythonPackages) != 2 || pythonPackages[0] != "polars==1.9.0" {
			t.Errorf("unexpected python_packages: %v", body["python_packages"])
		}
		writeJSON(t, w, 202, `{
			"id": "img-1", "label": "data-eng", "base": "boxkite-minimal",
			"python_packages": ["polars==1.9.0", "duckdb==1.1.3"], "apt_packages": [], "npm_packages": [],
			"status": "queued", "digest": null, "registry_ref": null, "scan_result": null,
			"failure_reason": null, "created_at": "2026-01-01T00:00:00Z", "completed_at": null
		}`)
	})
	defer closeServer()

	image, err := client.CreateImage(context.Background(), CreateImageRequest{
		Label:          Ptr("data-eng"),
		Base:           "boxkite-minimal",
		PythonPackages: []string{"polars==1.9.0", "duckdb==1.1.3"},
	})
	if err != nil {
		t.Fatalf("CreateImage: %v", err)
	}
	if image.ID != "img-1" || image.Status != "queued" {
		t.Errorf("unexpected image: %+v", image)
	}
}

func TestCreateImage_FeatureDisabledMapsToAPIError(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, 404, `{"error": {"code": "feature_disabled", "message": "Image builder not enabled."}}`)
	})
	defer closeServer()

	_, err := client.CreateImage(context.Background(), CreateImageRequest{})
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T", err)
	}
	if apiErr.Code != "feature_disabled" {
		t.Errorf("unexpected code: %s", apiErr.Code)
	}
}

func TestGetImage(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/images/img-1" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{
			"id": "img-1", "label": null, "base": "boxkite-default", "python_packages": [], "apt_packages": [], "npm_packages": [],
			"status": "completed", "digest": "sha256:abc", "registry_ref": "registry.internal/img-1@sha256:abc",
			"scan_result": {"critical": 0, "high": 0}, "failure_reason": null,
			"created_at": "2026-01-01T00:00:00Z", "completed_at": "2026-01-01T00:05:00Z"
		}`)
	})
	defer closeServer()

	image, err := client.GetImage(context.Background(), "img-1")
	if err != nil {
		t.Fatalf("GetImage: %v", err)
	}
	if image.Status != "completed" || image.Digest == nil || *image.Digest != "sha256:abc" {
		t.Errorf("unexpected image: %+v", image)
	}
}

func TestListImages(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, 200, `[]`)
	})
	defer closeServer()

	images, err := client.ListImages(context.Background())
	if err != nil {
		t.Fatalf("ListImages: %v", err)
	}
	if images == nil || len(images) != 0 {
		t.Errorf("expected empty non-nil slice, got %+v", images)
	}
}

func TestDeleteImage(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete || r.URL.Path != "/v1/images/img-1" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(204)
	})
	defer closeServer()

	if err := client.DeleteImage(context.Background(), "img-1"); err != nil {
		t.Fatalf("DeleteImage: %v", err)
	}
}

File: sdk-rust/tests/images_test.rs
mod common;

use boxkite_client::{CreateImageOptions, ImageBase};
use serde_json::json;
use wiremock::matchers::{body_json, method, path};
use wiremock::{Mock, ResponseTemplate};

#[tokio::test]
async fn create_image_sends_base_and_packages() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/images"))
        .and(body_json(json!({
            "label": "data-science",
            "base": "boxkite-default",
            "python_packages": ["polars==1.9.0"],
            "apt_packages": ["ripgrep==14.1.0-1"]
        })))
        .respond_with(ResponseTemplate::new(202).set_body_json(json!({
            "id": "img_1", "label": "data-science", "status": "queued", "created_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let options = CreateImageOptions::new()
        .label("data-science")
        .base(ImageBase::BoxkiteDefault)
        .python_packages(["polars==1.9.0"])
        .apt_packages(["ripgrep==14.1.0-1"]);

    let image = client
        .create_image(options)
        .await
        .expect("create_image should succeed");
    assert_eq!(image.id, "img_1");
    assert_eq!(image.status, "queued");
    // Fields absent from the "accepted" response default rather than erroring.
    assert!(image.python_packages.is_empty());
}

#[tokio::test]
async fn create_image_node_base_serializes_as_boxkite_node() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/images"))
        .and(body_json(
            json!({"base": "boxkite-node", "npm_packages": ["typescript==5.6.0"]}),
        ))
        .respond_with(ResponseTemplate::new(202).set_body_json(json!({
            "id": "img_2", "status": "queued", "created_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let options = CreateImageOptions::new()
        .base(ImageBase::BoxkiteNode)
        .npm_packages(["typescript==5.6.0"]);
    client
        .create_image(options)
        .await
        .expect("create_image should succeed");
}

#[tokio::test]
async fn get_image_parses_full_detail() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/images/img_1"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "id": "img_1", "label": "data-science", "base": "boxkite-default",
            "python_packages": ["polars==1.9.0"], "apt_packages": [], "npm_packages": [],
            "status": "completed", "digest": "sha256:abc", "registry_ref": "registry/img:abc",
            "scan_result": {"clean": true}, "failure_reason": null,
            "created_at": "2026-01-01T00:00:00Z", "completed_at": "2026-01-01T00:05:00Z"
        })))
        .mount(&server)
        .await;

    let image = client
        .get_image("img_1")
        .await
        .expect("get_image should succeed");
    assert_eq!(image.status, "completed");
    assert_eq!(image.digest.as_deref(), Some("sha256:abc"));
    assert_eq!(image.python_packages, vec!["polars==1.9.0"]);
}

#[tokio::test]
async fn list_images_returns_empty_vec_on_no_content() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/images"))
        .respond_with(ResponseTemplate::new(200))
        .mount(&server)
        .await;

    let images = client
        .list_images()
        .await
        .expect("list_images should succeed");
    assert!(images.is_empty());
}

#[tokio::test]
async fn delete_image_sends_delete() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("DELETE"))
        .and(path("/v1/images/img_1"))
        .respond_with(ResponseTemplate::new(204))
        .mount(&server)
        .await;

    client
        .delete_image("img_1")
        .await
        .expect("delete_image should succeed");
}

#[tokio::test]
async fn create_image_maps_feature_disabled_404() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/images"))
        .respond_with(ResponseTemplate::new(404).set_body_json(json!({
            "error": {"code": "feature_disabled", "message": "Custom images are not enabled on this deployment."}
        })))
        .mount(&server)
        .await;

    let err = client
        .create_image(CreateImageOptions::new())
        .await
        .unwrap_err();
    assert_eq!(err.code(), Some("feature_disabled"));
}
